torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -8,11 +8,69 @@
 #include <c10d/ProcessGroup.hpp>
 #include <c10d/Store.hpp>
 
+#include <ATen/Parallel.h>
 #include <ATen/cuda/CUDAContext.h>
 #include <ATen/cuda/CUDAEvent.h>
 
 namespace c10d {
 
+// CheckFutureWork is used by a cudaStremCallback function called
+// ncclKernelCompletionCallback. The purpose of CheckFutureWork is to mark
+// workNCCL's Future work when, all NCCL streams created in collective operation
+// are finished.
+// We use enable_shared_from_this to safely call shared_ptr<CheckFutureWork>
+// (this) to remove its pointer from ProcessGroupNCCL's checkFutObjs after
+// at::launch operation of markFutureCompleted is done.
+// markFutureCompleted increments streamCounter by one and once it is called by
+// number of replicas times, it marks Future work completed by setting outputs
+// vector as its value.
+struct CheckFutureWork : std::enable_shared_from_this<CheckFutureWork> {
+ public:
+  CheckFutureWork(
+      std::shared_ptr<ProcessGroup::Work> work,
+      std::vector<at::Tensor>& outputs,
+      std::shared_ptr<std::unordered_set<std::shared_ptr<CheckFutureWork>>>
+          checkFutObjs,
+      std::mutex& checkFutObjMutex_)
+      : work_(work),
+        outputs_(std::make_shared<std::vector<at::Tensor>>(outputs)),
+        streamCounter_(0),
+        checkFutObjs_(checkFutObjs),
+        checkFutObjMutex_(checkFutObjMutex_){};
+
+  std::shared_ptr<CheckFutureWork> getPtr() {
+    return shared_from_this();
+  }
+
+  void markFutureCompleted() {
+    at::launch(([this]() {
+      std::unique_lock<std::mutex> lock(checkFutObjMutex_);
+      (streamCounter_)++;
+      if (streamCounter_ == (*outputs_).size()) {
+        // Need to wait until work is completed, because some post processing
+        // could be needed for outputs.
+        work_->wait();
```

Comment:
I think we might be able to get rid of this if we consider the group-guard related change I mentioned below.

Reply:
replaced with `synchronizeStreams`